### PR TITLE
Support `Serialize` + `Deserialize` for `dashu-ratio`

### PR DESCRIPTION
**Pull request by oovm**
- string form

```json
"-5/11"
```

- object form

```json
{
    "sign": true,
    "numerator": [5],
    "denominator": [11]
}
```

### CODE REVIEW

File: rational/src/third_party/ser.rs [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+use crate::RBig;
+
+impl Serialize for RBig {
+    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
+        if serializer.is_human_readable() {
+            serializer.serialize_str(&self.to_string())
+        } else {
+            let sign = self.sign();
+            let numerator = self.numerator().clone().into_parts().1.to_le_bytes();
```
Comment (oovm) [inline rational/src/third_party/ser.rs:14]: There is one more clone here, because the binary form cannot be obtained directly

**Comment by oovm** [comment]
Well, sorry, seems not work with `no-std`

**Comment by cmpute** [comment]
Thanks for your contribution, but the implementation seems over-complicate for me. Do you have any specific reason to serialize the rational number as a struct with three parts? I was planning to serialize the rational number just as a pair of integers.

Besides, you can take a look at #11, where I listed all thirdparty crates I planned to support, so you don't have to implement them yourselves.

**Comment by oovm** [comment]
How about dividing into the following three binary representations?

1. `[ ]` => 0 / 1
2. `[i64, bytes<numerator>]` => numerator / 1
3. `[i64, bytes<numerator>, u64, bytes<denominator>]` => numerator / denominator

**Comment by oovm** [comment]
Maybe [VLQ](https://en.wikipedia.org/wiki/Variable-length_quantity) is better than `i64` and `u64`, since it also supports non-64-bit platforms.

**Comment by cmpute** [comment]
I don't think we need to handle the VLQ implementation ourselves, it could be handled by serializers like the `postcard`. I already have `serde` support for `UBig` and `IBig`, we should reuse the implementation of that. (Although I am planning to refactor it to use byte array instead of the word array which is currently used).

What I have planned for the (binary) serialization formats is:
- `UBig` serialized as a byte array (with given length)
- `IBig` serialized as a byte array with the sign stored in the last byte
- `RBig` serialized as a (named) tuple of `(UBig, IBig)`
- `FBig` serialized as a (named) tuple of `(IBig, isize, usize)`

For text serialization, I would just use the format result of each type.

**Comment by cmpute** [comment]
`serde` support for `dashu-float` and `dashu-ratio` has been added in https://github.com/cmpute/dashu/commit/d184f7cbb7654ccad692d7ff14d88c8e9f333e67, the improvements on serializing `UBig` and `IBig` will be added after releasing the next minor version, because that is a break change.

Thanks for your help but I will close this issue for now.

**Pull request closed without merging**